### PR TITLE
`deprecated` attribute support for notations and syntactic definitions

### LINKED ISSUE #6840: Compatibility notations do not have the same semantics as normal notations (or only parsing notations).
#### Version

8.8+alpha (current master)

#### Description of the problem

```coq
Notation "[ x ]" := (cons x nil) : list_scope.
Notation "[ x ; .. ; y ]" := (cons x .. (cons y nil) ..) (compat "8.6") : list_scope.

Variable Foo : nat -> nat.

Delimit Scope Foo_scope with F.

Notation " [ x ] " := (Foo x) : Foo_scope.

Check ([1] : nat)%F.
```

works.

```coq
Notation "[ x ]" := (cons x nil) : list_scope.
Notation "[ x ; .. ; y ]" := (cons x .. (cons y nil) ..) : list_scope.

Variable Foo : nat -> nat.

Delimit Scope Foo_scope with F.

Notation " [ x ] " := (Foo x) : Foo_scope.

Check ([1] : nat)%F.
```

doesn't.

```coq
Notation "[ x ]" := (cons x nil) : list_scope.
Notation "[ x ; .. ; y ]" := (cons x .. (cons y nil) ..) (only parsing) : list_scope.

Variable Foo : nat -> nat.

Delimit Scope Foo_scope with F.

Notation " [ x ] " := (Foo x) : Foo_scope.

Check ([1] : nat)%F.
```

doesn't either.

I thought the compatibility annotation was just about emitting warnings but apparently it changes also the semantics. I don't know if that's wanted or not.

### PR DESCRIPTION
**Pull request by maximedenes**
We also slightly change the semantics of the `compat` syntax modifier to
re-express it in terms of the `deprecated` attribute, and we deprecate
it in favor of the latter.

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

Fixes #6840.

### CODE REVIEW

File: theories/Logic/Berardi.v
Comment (SkySkimmer) [inline theories/Logic/Berardi.v]: Isn't this going to say "foo is deprecated, use foo instead"?

Comment (maximedenes) [reply theories/Logic/Berardi.v]: I don't think so :)

File: vernac/vernacentries.ml
Comment (SkySkimmer) [inline vernac/vernacentries.ml]: Would it make sense to define a `module_locality = locality >>= fun l -> return (make_module_locality l)` attribute?

File: interp/deprecation.mli
Comment (SkySkimmer) [inline interp/deprecation.mli]: Should we provide a `val pr : t -> Pp.t`?

File: theories/Logic/Berardi.v
Comment (SkySkimmer) [reply theories/Logic/Berardi.v]: oh it say "fou is deprecated" tricky

File: doc/sphinx/language/gallina-specification-language.rst
Comment (herbelin) [inline doc/sphinx/language/gallina-specification-language.rst]: It could be more explicit, telling which `@string` is which between `since` and `note`. It is probably worth also to give the syntax for the version number (use of dash, of dots, prefxed with a `V` or not, etc.).

**Review by herbelin — CHANGES_REQUESTED** [review]
It is still a mystery to me how one could prefer a syntax which sounds to me like arid and "shell"-looking to a smoother syntax such as `deprecated since V8.3 using message "foo"`, but this PR is otherwise ok to me (up to the documentation which is certainly worth being made more precise).

File: doc/sphinx/language/gallina-specification-language.rst
Comment (Zimmi48) [reply doc/sphinx/language/gallina-specification-language.rst]: I wouldn't constrain the syntax for the version number given that if people use it for their own libraries, it could be "is deprecated since My Awesome Library 0.9" or whatever they want to.

Here is a suggestion for the more explicit distinction:

```rst
.. warn:: Notation @string__1 is deprecated since @string__2. @string__3

   :n:`@string__1` is the actual notation, :n:`@string__2` is the version number,
   :n:`@string__3` is the note.
```

**Review by Zimmi48 — COMMENTED** [review]
Finally, we get rid of these compatibility notations! :tada: (cc #6840 which can be closed when this is the case)

Indeed, I agree the original documentation of the `deprecated` attribute was quite limited and it should be improved in passing. In particular, it should mention to which command the attribute can be attached. It used to be `Ltac` and `Tactic Notation` only, and now `Notation` is supported as well (but only if it is an abbreviation?). BTW, what about `Ltac2`?

File: doc/sphinx/language/gallina-specification-language.rst
Comment (herbelin) [reply doc/sphinx/language/gallina-specification-language.rst]: The suggestion for `string` is good. For the version number, the ability to use it in a library is indeed relevant. I'm still afraid though that we introduce inconsistency within the code that we manage, if we don't write a rule somewhere telling how it should be. Remember also that there is an option `-compat X.Y` (parsed in `G_vernac.parse_compat_version`) and that this version is legitimatedly supposed to have an action on the activation of the deprecation flag. So, maybe can we have two syntaxes for the version number, one unconstrained for libraries, and one constrained so as to work in agreement with the `-compat` option?

Comment (maximedenes) [reply doc/sphinx/language/gallina-specification-language.rst]: I deliberately changed the semantics of compatibility notations, as I think they made little sense. Tying it to Coq versions is very strange, as is their activation depending on `-compat`. I think the warning-based approach is more sensible, especially for library authors.

The only part I kept (and I'm not fully sure about) is that deprecated notations become parsing only, so that the printer does not pick up deprecated notations. @ggonthier do you think that's the way it should be?

Comment (Zimmi48) [reply doc/sphinx/language/gallina-specification-language.rst]: I fully support removing the link between the `-compat` option and deprecated compatibility notations. This is what I was complaining about in #6840 and that's why I'm saying this PR fixes this bug.

Comment (herbelin) [reply doc/sphinx/language/gallina-specification-language.rst]: I'm a bit lost. Do I understand correctly that the complaint in #6840 was about the line `if is_active_compat se.synext_compat` in `metasyntax.ml`, i.e. about the semantics change induced by `-compat`? What I was talking about above was instead about the activation of the deprecation warning. Wouldn't that be a bit strange if you use `-compat 8.9` and get a warning that it is deprecated since 8.10? Anyway, that's not a big deal, and it can be amended in further PRs, so do your best.

About deactivating the printing when deprecated, why not. By the way, what do you intend to do about printing when the warning is silenced?

Comment (ggonthier) [reply doc/sphinx/language/gallina-specification-language.rst]: > The only part I kept (and I'm not fully sure about) is that deprecated notations become parsing only, so that the printer does not pick up deprecated notations. @ggonthier do you think that's the way it should be?

I think so, although this is only meaningful for definition, since one rarely views proof terms.
One feature I think would be useful, perhaps even more so than the `since` argument, would be the ability to specify a warning category, so that the warnings related to a particular library, and perhaps a particular update of a library, could be easily selected.

Comment (Zimmi48) [reply doc/sphinx/language/gallina-specification-language.rst]: > Wouldn't that be a bit strange if you use `-compat 8.9` and get a warning that it is deprecated since 8.10?

I don't think so. The `-compat` mechanism should be kept as simple as possible IMHO and be limited to the loading of the appropriate `Compat.v` file.

Comment (herbelin) [reply doc/sphinx/language/gallina-specification-language.rst]: > The -compat mechanism should be kept as simple as possible IMHO and be limited to the loading of the appropriate `Compat.v` file.

Ok, that's defensible.

Comment (maximedenes) [reply doc/sphinx/language/gallina-specification-language.rst]: Suggestion applied, thanks.

File: interp/deprecation.mli
Comment (maximedenes) [reply interp/deprecation.mli]: I think we can refine the API later, if that's ok.

**Review by herbelin — APPROVED** [review]

File: theories/Logic/Berardi.v
Comment (JasonGross) [reply theories/Logic/Berardi.v]: Also, having this on a separate line breaks the ability of the python script I wrote to automatically remove this.

File: test-suite/bugs/closed/bug_4798.v
Comment (JasonGross) [inline test-suite/bugs/closed/bug_4798.v]: Why do you remove this file rather than adapting it to use a deprecated notation?

**Review by JasonGross — CHANGES_REQUESTED** [review]
I support this change in principle, but I want to know what the story is for updating versions before I approve; previously, my python script automatically removed outdated compat notations after two versions.  This seems like a nice feature.  What's the new policy on how long we keep around deprecated notations, and what tooling support do we want for it?

File: theories/Logic/Berardi.v
Comment (maximedenes) [reply theories/Logic/Berardi.v]: Like I wrote after @SkySkimmer's first review, the name is not the same. Basically this was fixing a typo.

@Zimmi48 what do you think I should do about the script?

File: interp/deprecation.ml
Comment (maximedenes) [reply interp/deprecation.ml]: This is the case in other files, I don't know what the story is for these dates. @Zimmi48 ?

File: test-suite/bugs/closed/bug_4798.v
Comment (maximedenes) [reply test-suite/bugs/closed/bug_4798.v]: IIUC, this file checks that `compat` notations are inactive when the corresponding `-compat` flag is not passed. This is no longer the case (see changelog), so I had to remove the test.

File: dev/tools/update-compat.py
Comment (maximedenes) [reply dev/tools/update-compat.py]: Done, thanks.

**Comment by maximedenes** [comment]
Anybody who can self-assign? @Zimmi48 ? @herbelin ?

File: theories/Logic/Berardi.v
Comment (Zimmi48) [reply theories/Logic/Berardi.v]: I think that this script was mostly important for the large number of deprecated notations removal that we got in the last release, but it would be nice if we could keep it working. @JasonGross isn't the two-line problem something that could be fixed in the script? It seems OK to put the fully qualified name in the error message if that can help.

File: interp/deprecation.ml
Comment (Zimmi48) [reply interp/deprecation.ml]: Please update it to 2019. See #10320 for the other files.

File: vernac/metasyntax.ml
Comment (Zimmi48) [inline vernac/metasyntax.ml]: Previously, the warning was:

```
Notation foo := plus (compat "8.8").
Check foo. (* foo is Nat.add [compatibility-notation,deprecated] *)
```

Now it will be `Notation foo is deprecated since 8.8.`

Keeping the warning message unchanged is going to be unnecessarily tedious but at least, it should include a note giving the qualid to replace the notation.

**Review by Zimmi48 — COMMENTED** [review]
The only real issue is with this regression in the content of the user message.

File: theories/Logic/Berardi.v
Comment (maximedenes) [reply theories/Logic/Berardi.v]: The name wasn't qualified already with `compat`, so I'll keep it as it was in this PR.

File: vernac/metasyntax.ml
Comment (maximedenes) [reply vernac/metasyntax.ml]: This is going to require significant changes in the API, just for the compat layer. Who's using `compat`?

**Review by Zimmi48 — APPROVED** [review]
Fine.

**Comment by Zimmi48** [comment]
>I want to know what the story is for updating versions before I approve; previously, my python script automatically removed outdated compat notations after two versions. This seems like a nice feature. What's the new policy on how long we keep around deprecated notations, and what tooling support do we want for it?

I think we want to keep the same policy, but this change means that we go back to the previous situation where deprecated notations are removed manually (like deprecated options, commands, etc.). This means that we are more flexible, which is actually a good thing. If for some reason, one of our deprecated notation has to be kept for longer because of a specific user requesting more time to port their project, we can delay its removal (as we do already for deprecated options, commands, etc.) without changing the "since" sub-attribute, because the deprecated notation infrastructure is no longer linked to the compat infrastructure.

**Comment by Zimmi48** [comment]
@JasonGross Can you approve the PR if you are satisfied by the changes and answers?

**Comment by Zimmi48** [comment]
ping @JasonGross 
Will merge tomorrow if no answer.

**Comment by maximedenes** [comment]
@Zimmi48 ping

**Comment by Zimmi48** [comment]
I am assuming the QuickChick failure is unrelated.

**Comment by maximedenes** [comment]
> I am assuming the QuickChick failure is unrelated.

AFAIK it was occurring on master at the time this PR was submitted.

**Comment by ejgallego** [comment]
I'd like to remark that math-comp introduced support for deprecating definitions using a notation https://github.com/math-comp/math-comp/pull/342 , I do indeed wonder how the attribute-based system and the math-comp approach do relate.

**Comment by ggonthier** [comment]
The MathComp notation is mostly a shim: we  had an immediate need for supporting compatibility notation, and won't be able to use anything Coq introduces for at least a year because we strictly support 2-version backward compatibility.
  The Notation approach has several drawbacks: the warnings are not true warnings so can't be managed with options (we have to use Canonical instances for that!), the warnings are emitted during type inference leading to multiple warnings when type inference is performed multiple times (e.g., for views), and the compatibility notations can only approximate the implicit arguments of the originals because Coq only supports `Arguments` signature for pure alias `Notation` (and even then, we have to work around an ugly context pruning incompleteness in unification). Definitely, attributes are a better way to go.

**Pull request merged**
